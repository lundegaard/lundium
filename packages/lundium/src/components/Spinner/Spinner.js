import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { cx } from 'ramda-extension';

import Text from '../Text';
import Grid from '../Grid';

const Spinner = ({
	size = 'lg',
	visibilityDelay = 3000,
	animationDisabled,
	children,
	className,
}) => {
	const [showLoader, setShowLoader] = useState(visibilityDelay <= 0);

	useEffect(() => {
		const timer = setTimeout(changeVisibility, visibilityDelay);

		return () => {
			clearTimeout(timer);
		};
	}, [visibilityDelay]);

	const changeVisibility = () => setShowLoader(true);

	const isSmall = size === 'sm';
	const gridClasses = cx([
		'data-status',
		...(animationDisabled ? [] : ['fade-in']),
	]);

	if (!showLoader) {
		return null;
	}

	const svg = (
		<svg
			className={cx(className, 'loader', {
				'loader--sm': isSmall,
			})}
			viewBox="0 0 60 60"
			xmlns="http://www.w3.org/2000/svg"
		>
			<circle
				className="loader-circle"
				cx="30"
				cy="30"
				r={isSmall ? '25' : '20'}
				stroke="#70c542"
				strokeWidth={isSmall ? '5' : '3'}
			/>
		</svg>
	);

	if (!children) {
		return svg;
	}

	return (
		<Grid alignItems="center" justifyContent="center" className={gridClasses}>
			{svg}
			<Text mb={0}>{children}</Text>
		</Grid>
	);
};

Spinner.propTypes = {
	/** Determines wheter to animate or not. */
	animationDisabled: PropTypes.bool,
	children: PropTypes.node,
	className: PropTypes.string,
	size: PropTypes.oneOf(['sm', 'lg']),
	/** Delay for showing loader in ms. */
	visibilityDelay: PropTypes.number,
};

export default Spinner;
