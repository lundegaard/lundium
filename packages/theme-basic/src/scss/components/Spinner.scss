@import '../core';

/*
 -  Loading spinner
 -  (<div class="loader" />)
 */


.loader {
	width: 60px;
	height: 60px;
	margin-inline-end: $spacer;
	stroke-linecap: round;
	stroke-width: 6;
	fill: none;

	&--sm {
		height: 16px;
		width: 16px;
	}

	.loader-circle {
		stroke: $theme-blue;
		// stroke-dashoffset: 0;
		transform-origin: center;
		animation: loadingAnim 1.8s ease-in-out infinite;

		&:dir(rtl) {
			animation-direction: reverse;
		}
	}
}

@keyframes loadingAnim {
	0% {
		stroke-dasharray: 1, 200;
		stroke-dashoffset: 0;
		transform: rotate(0);
		stroke: $theme-blue;
	}
	50% {
		stroke-dasharray: 89, 200;
		stroke-dashoffset: -35;
		stroke: $theme-green;
	}
	100% {
		stroke-dasharray: 89, 200;
		stroke-dashoffset: -124;
		transform: rotate(360deg);
		stroke: $theme-blue;
	}
}

.fade-in {
	animation: fadeIn 2s;
	animation-fill-mode: forwards;
}

@keyframes fadeIn {
	from {
		opacity: 0;
	}
	to {
		opacity: 1;
	}
}
