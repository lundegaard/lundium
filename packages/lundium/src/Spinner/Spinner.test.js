import React from 'react';
import { shallow } from 'enzyme';

import Spinner from './Spinner';

describe('<Spinner />', () => {
	const requiredProps = {
		animationDisabled: true,
		visibilityDelay: 0,
		children: <div>Loading data</div>,
	};

	it('matches snapshot when shallowly rendered', () => {
		const wrapper = shallow(<Spinner {...requiredProps} />);
		expect(wrapper).toMatchSnapshot();
	});
});
