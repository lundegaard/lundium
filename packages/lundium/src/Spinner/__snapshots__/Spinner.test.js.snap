// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Spinner /> matches snapshot when shallowly rendered 1`] = `
<forwardRef(Grid)
  alignItems="center"
  className="data-status"
  justifyContent="center"
>
  <svg
    className="loader"
    viewBox="0 0 60 60"
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      className="loader-circle"
      cx="30"
      cy="30"
      r="20"
      stroke="#70c542"
      strokeWidth="3"
    />
  </svg>
  <forwardRef(Text)
    mb={0}
  >
    <div>
      Loading data
    </div>
  </forwardRef(Text)>
</forwardRef(Grid)>
`;
